README.md
# OECD Simple Charts

A lightweight chartingtool. Visit the [docs](https://oecd-cyc.github.io/oecd-simple-charts) or checkout the [code](https://github.com/oecd-cyc/oecd-simple-charts).

<img src="https://raw.githubusercontent.com/oecd-cyc/oecd-simple-charts/master/public/images/charts-example.png" style="width:100%; max-width: 960px">

## Usage

To get started, you can either install the library using npm

```
npm install oecd-simple-charts
```

or embed the `bundle.min.js` and `bundle.min.css` files, either downloading them from the [repository](https://github.com/oecd-cyc/oecd-simple-charts) or directly embedding them from a CDN service like [unpkg](https://unpkg.com/oecd-simple-charts/build/oecd-simple-charts.min.js) or [jsDelivr](https://unpkg.com/oecd-simple-charts/build/oecd-simple-charts.min.js).

```html
<script src="https://unpkg.com/oecd-simple-charts/build/oecd-simple-charts.min.js"></script>
<link href="https://unpkg.com/oecd-simple-charts/build/oecd-simple-charts.min.css" rel="stylesheet">
```

In order to create a chart you need to add a container DOM node. Then you can start with the configuration of the chart. You can set a title, change the size and color of the elements and add data points. Each chart has an `update` function, that takes an array of new data and updates the visualization.

## Development

Clone the repository, install the dependencies, builds the application and starts a webserver with hot loading on [localhost:3000](http://localhost:3000/):

```sh
$ git clone https://github.com/oecd-cyc/oecd-simple-charts.git
$ cd oecd-simple-charts
$ npm install
$ npm start
```

## Build

Builds a minified version of the library in the `build/` folder.

```sh
$ npm run build
```

## Testing

Run the functional test-suite located in `test/`:

```sh
$ npm test
```

To run the visual-regression tests, you will first need a running webserver on port 3000 (see [Development](#development)). Then you can start the tests running:

```sh
$ npm run test:visual
```

If the test fails, but you want to keep your changes run:

```sh
$ npm run test:visual:reference
```

This will save new screenshots in the `screenshots/` folder and these will be used as reference to be tested against the next time you run the tests.

## Update documentation

In order to generate the docs, run:

```sh
$ npm run docs
```


CHANGELOG.md


# OECD Charting-tool

  

  _Recent changes:_
  




  ## v1.0.0
  
  




## Bug Fixes

  - **box plot**
    - full width for charts, add titles to all chartsections
  ([f2a0cd15](https://github.com/wbkd/oecd_charting-tool/commit/f2a0cd154f981747aeb2e503a99f3e34a3ef3c6b))

  - **charts**
    - firefox baseline alignment
  ([a7a76655](https://github.com/wbkd/oecd_charting-tool/commit/a7a76655998acd080b4e38b82f8c17d1dc669360))

  - **docs**
    - add examples for the PearlChart
  ([ec2c3cf8](https://github.com/wbkd/oecd_charting-tool/commit/ec2c3cf84b980e265652227b5b7aa6bcee58e765))

  - **package**
    - mark package.json as a binary file
  ([5d1823a3](https://github.com/wbkd/oecd_charting-tool/commit/5d1823a3c3f88ce7e5c4499863921363e3138623))
    - take my package-lock.json
  ([d0730717](https://github.com/wbkd/oecd_charting-tool/commit/d073071755e6c1be580a454b9b1701f80794f062))
    - mark package.json as a binary file
  ([a5a617bb](https://github.com/wbkd/oecd_charting-tool/commit/a5a617bbfc3dd6338152a50007fcd016cbb688a9))

  - **test**
    - use d3.select for all tests
  ([322818fb](https://github.com/wbkd/oecd_charting-tool/commit/322818fb1fe59af0cc90a275db6313dc9e44fc66))
    - use d3.select() and update PearlChartTest
  ([db019486](https://github.com/wbkd/oecd_charting-tool/commit/db0194868f63882c6a26a367a20e3451d30e199d))




## Features
  - add labels and use scales for bars
  ([d946b306](https://github.com/wbkd/oecd_charting-tool/commit/d946b3062a1638f9d29ba375de93757bd35d8afe))
  - basic stacked bar charts
  ([b8ce13e7](https://github.com/wbkd/oecd_charting-tool/commit/b8ce13e7bd765c585a867a7e3d40f1f835d059e4))

  - **box plot**
    - add labels
  ([dc2151a4](https://github.com/wbkd/oecd_charting-tool/commit/dc2151a45702733a5c58dc2064e186a60c9670ab))
    - set height depending on data
  ([75d05251](https://github.com/wbkd/oecd_charting-tool/commit/75d052517e995ebba5b4a1d8712e05f4c4deb664))
    - display markers
  ([ff5bf68e](https://github.com/wbkd/oecd_charting-tool/commit/ff5bf68ec8df64559e0f7cffdebf0cafdface405))
    - display axis and grid
  ([9b877808](https://github.com/wbkd/oecd_charting-tool/commit/9b87780842e42ca0f45061d526ab27a8a27eb1d3))

  - **changelog**
    - autogenerate CHANGELOG.md
  ([d2c40b23](https://github.com/wbkd/oecd_charting-tool/commit/d2c40b231bde687690270854458a737ad777c08e))

  - **docs**
    - update the README and add static docs generator
  ([4f585f5d](https://github.com/wbkd/oecd_charting-tool/commit/4f585f5dffeb23e3f853ef21a3b1e70969ce0076))

  - **pearlchart**
    - setup pearl chart module
  ([d1852ea2](https://github.com/wbkd/oecd_charting-tool/commit/d1852ea2d20253b7c1a7d13fb53edeff4b588a3b))

  - **stacked bar charts**
    - check if labels exist
  ([ba30590a](https://github.com/wbkd/oecd_charting-tool/commit/ba30590ab739de139d25044e7cb77f0f75cf81f8))

  - **test**
    - add visual regression testing
  ([5b4ded68](https://github.com/wbkd/oecd_charting-tool/commit/5b4ded68d2bb6ccf905a382be9a0b50227703a27))
    - add tests
  ([0d5bc045](https://github.com/wbkd/oecd_charting-tool/commit/0d5bc04566a6fbd4921e4e8397d7ee5a42acf5f8))




## Documentation

  - **jsdoc**
    - update docs
  ([2e03b823](https://github.com/wbkd/oecd_charting-tool/commit/2e03b823bb19ed3f23a25aed4a56bdbe1aa28a00))
    - disable timestamp in generated docs
  ([8dbbe080](https://github.com/wbkd/oecd_charting-tool/commit/8dbbe08025379043c8d3e0a66fb8d832615de933))
    - update docs
  ([70302cff](https://github.com/wbkd/oecd_charting-tool/commit/70302cff054897ffc236615011f7f5896023bd27))
    - partially revert the last commit
  ([6f077954](https://github.com/wbkd/oecd_charting-tool/commit/6f0779543a164b82030b9bb1533e6836ce200450))
    - update docs
  ([2ea07b41](https://github.com/wbkd/oecd_charting-tool/commit/2ea07b41fc54c7624fb3b1499bd16bbd73891922))
    - use @example tag instead of markdown
  ([bf61fb64](https://github.com/wbkd/oecd_charting-tool/commit/bf61fb64db76b8766c38550fa7b9985179a9c03c))
    - update docs
  ([60c917f2](https://github.com/wbkd/oecd_charting-tool/commit/60c917f280ef250a1a1211ccbef925d8b5e1de51))
    - add some examples
  ([95b56c84](https://github.com/wbkd/oecd_charting-tool/commit/95b56c84b3638d98d23d6227af24736127a43096))
    - add jsdoc
  ([34c3a124](https://github.com/wbkd/oecd_charting-tool/commit/34c3a124fac6021faa2eb945bd68d7b630da8c68))
    - add some examples
  ([1c68724a](https://github.com/wbkd/oecd_charting-tool/commit/1c68724a0a1e3b4aea62f7cb788e390b331a4ef2))
    - add jsdoc
  ([993a696d](https://github.com/wbkd/oecd_charting-tool/commit/993a696d5049f23106518a2df4aa2b8fcf9f850e))

  - **package**
    - update package.json meta
  ([55f8f205](https://github.com/wbkd/oecd_charting-tool/commit/55f8f205933aa9b8c5b80f0aea94ca9f2d8793bd))

  - **readme**
    - add docs instructions to the README
  ([e740423a](https://github.com/wbkd/oecd_charting-tool/commit/e740423aeadb7c9e71b9b066cc065e5bb36ef7c9))
    - add a screenshot to the readme
  ([b430012c](https://github.com/wbkd/oecd_charting-tool/commit/b430012c7559637f6e8113c2b65ec1087cae9709))
    - update the README
  ([ef4559ca](https://github.com/wbkd/oecd_charting-tool/commit/ef4559ca8fc5ffe7759abf8110960683cd944afd))
    - update the README
  ([811bfebe](https://github.com/wbkd/oecd_charting-tool/commit/811bfebef8c673881763d42bef51328188e90537))

  - **screenshot**
    - add screenshot to the readme/docs
  ([aaf08604](https://github.com/wbkd/oecd_charting-tool/commit/aaf0860465812e8a56cb15733bf2147d0cf76b14))

  - **screenshots**
    - update visual testing reference and screenshots
  ([8f7ab7ae](https://github.com/wbkd/oecd_charting-tool/commit/8f7ab7aee309a3d891c62ca08c9998354713bec9))




## Refactor
  - rename css classes, fix eslint-warnings
  ([0ca24c90](https://github.com/wbkd/oecd_charting-tool/commit/0ca24c9055d3a201039c6607fe613e7887eb65c3))

  - **package**
    - refactor npm scripts

test/BoxPlotTest.js
import test from 'ava';
import { select as d3Select } from 'd3-selection';
import { dispatch as d3Dispatch } from 'd3-dispatch';

import BoxPlot from '../src/modules/BoxPlot';

const container = document.createElement('div');
container.id = 'BoxPlotExample';
document.body.appendChild(container);

const options = {
  container: '#BoxPlotExample',
  title: 'Average Peformance',
  extent: [350, 650],
  step: 50,
  legend: '<div class="legend">Legend</div>',
  renderInfoButton: true,
  data: [
    {
      values: [480, 500, 530],
      color: ['#f0f0f0', '#555', '#000'],
      labelLeft: '<img src="path/to.svg"/>',
      labelRight: 'Some text'
    },
    {
      values: [380, 450, 500],
      color: ['#f0f0f0', '#555', '#000'],
      labelLeft: '<img src="path/to.svg"/>',
      labelRight: 'Some text'
    },
    {
      values: [400, 520, 570],
      color: ['#EBCFCC', '#D14432', '#D14533']
    },
    {
      values: [410, 520, 560],
      color: ['#EBCFCC', '#D14432', '#D14533']
    }
  ]
};

const boxPlot = new BoxPlot(options);

const containerSelection = d3Select('.OECDCharts__BoxPlot');
test('has a .OECDCharts__BoxPlot class name', (t) => {
  t.true(!!containerSelection.size());
});

test('has an update function', (t) => {
  t.true(typeof boxPlot.update === 'function');
});

test('has a listener for infoclick', (t) => {
  const infoIcon = containerSelection.select('.oecd-chart__info');
  boxPlot.on('infoclick', () => {
    t.pass();
  });

  infoIcon.node().click();
});

test('displays a title', (t) => {
  const title = containerSelection.select('.oecd-chart__title');
  t.true(title.text() === options.title);
});

test('container contains an info icon', (t) => {
  const infoIcon = containerSelection.select('.oecd-chart__info-icon');
  t.true(!!infoIcon.size());
});

test('boxplot chart has legend', (t) => {
  const legend = containerSelection.select('.boxplot__legend');
  t.true(!!legend.size());
});

const svgContainer = containerSelection.select('.boxplot__svg');
test('contains an <svg> element with width and height', (t) => {
  t.true(!!svgContainer.attr('width') && !!svgContainer.attr('height'));
});

const g = svgContainer.select('g');
test('the <svg> element has <g> children', (t) => {
  t.true(!!g.size());
});

test('contains a markergroup', (t) => {
  const markerGroups = svgContainer.selectAll('.boxplot__marker-group');
  t.true(markerGroups.size() === 4);
});

// test('the <svg> element contains left labels for markergroups', (t) => {
//   const labelLeft = svgContainer.selectAll('.marker-group__label--left');
//   t.true(labelLeft.size() === 4);
// });

// test('the <svg> element contains right labels for markergroups', (t) => {
//   const labelRight = svgContainer.selectAll('.marker-group__label--right');
//   t.true(labelRight.size() === 4);
// });

const xAxis = svgContainer.select('.boxplot__axis');
test('xaxis has ticks depending on extend and stepwidth', (t) => {
  const expectedTicks = ((options.extent[1] - options.extent[0]) + options.step) / options.step;
  const tickNodes = xAxis.selectAll('.tick');
  t.true(tickNodes.size() === expectedTicks);
});

test('has a toggle title function', (t) => {
  t.true(typeof boxPlot.toggleTitle === 'function');
});



test/PearlChartTest.js
import test from 'ava';
import { select as d3Select } from 'd3-selection';

import PearlChart from '../src/modules/PearlChart';

const container = document.createElement('div');
container.id = 'PearlChartExample';
document.body.appendChild(container);

const ob = new PearlChart({
  container: '#PearlChartExample',
  extent: [300, 1000],
  renderInfoButton: true,
  data: [
    {
      value: 520,
      color: '#f00'
    },
    {
      value: 800
    }
  ]
});

const containerSelection = d3Select('.OECDCharts__PearlChart');
test('has a .OECDCharts__PearlChart container', (t) => {
  t.true(!!containerSelection.size());
});

test('has an update function', (t) => {
  t.true(typeof ob.update === 'function');
});

test('has a listener for infoclick', (t) => {
  const infoIcon = containerSelection.select('.oecd-chart__info');
  ob.on('infoclick', () => {
    t.pass();
  });

  infoIcon.node().click();
});

test('has an info-icon and a title', (t) => {
  const titleContainer = containerSelection.select('.oecd-chart__title-container');
  const infoIcon = titleContainer.select('.oecd-chart__info');
  t.true(!!titleContainer.select('.oecd-chart__title').size() && !!infoIcon.select('svg.oecd-chart__info-icon').select('path').size());
});

test('container contains an info icon', (t) => {
  const infoIcon = containerSelection.select('.oecd-chart__info-icon');
  t.true(!!infoIcon.size());
});

const svgContainer = containerSelection.select('svg.pearlchart__svg');
test('contains an <svg> element with width and height attributes', (t) => {
  t.true(!!svgContainer.attr('width') && !!svgContainer.attr('height'));
});

const g = svgContainer.select('g');
test('the <svg> element has <g> children', (t) => {
  t.true(!!g.size());
});

const xAxis = g.select('.pearlchart__x-axis');
test('the <g> element contaits the x-axis group', (t) => {
  t.true(!!xAxis.size());
});

test('the x-axis group contains a line element', (t) => {
  t.true(!!xAxis.select('line').size());
});

test('the x-axis group contains two <text> elements', (t) => {
  t.true(xAxis.selectAll('text').size() === 2);
});

test('the <text> elements has the class axis-label', (t) => {
  t.true(xAxis.selectAll('.pearlchart__axis-label').size() === 2);
});

const yAxis = g.select('.pearlchart__nodes');
test('the <g> element contains the nodes group', (t) => {
  t.true(!!yAxis.size());
});

test('the y-axis group contains two <circle> elements', (t) => {
  t.true(yAxis.selectAll('circle').size() === 2);
});

test('the y-axis group contains two <text> elements', (t) => {
  t.true(yAxis.selectAll('text').size() === 2);
});

test('has a toggle title function', (t) => {
  t.true(typeof ob.toggleTitle === 'function');
});

test/StackedChartTest.js
import test from 'ava';
import { select as d3Select } from 'd3-selection';

import StackedChart from '../src/modules/StackedChart';

const container = document.createElement('div');
container.id = 'StackedChartExample';
document.body.appendChild(container);

const options = {
  container: '#StackedChartExample',
  title: 'Distribution of performance level I - IV',
  renderInfoButton: false,
  data: [
    {
      values: [1, 2, 3, 4, 5],
      barLabels: ['0%', '100%'],
      colors: ['#D1CFCF', '#4C4C4C'],
      stackLabels: ['I', 'II', 'III', 'IV', 'V']
    },
    {
      values: [2, 4, 6, 8, 20],
      barLabels: ['0%', '100%'],
      colors: ['#fff', '#f00']
    }
  ]
};

const sc = new StackedChart(options);

test('has an update function', (t) => {
  t.true(typeof sc.update === 'function');
});

const containerSelection = d3Select('.OECDCharts__StackedChart');
test('has a .OECDCharts__StackedChart class name', (t) => {
  t.true(!!containerSelection.size());
});

test('displays a title', (t) => {
  const title = containerSelection.select('.oecd-chart__title');
  t.true(title.text() === options.title);
});

test('container contains no info icon', (t) => {
  const infoIcon = containerSelection.select('.oecd-chart__info-icon');
  t.true(!infoIcon.size());
});

const svgContainer = containerSelection.select('svg.stacked-chart__svg');
test('contains an <svg> element', (t) => {
  t.true(!!svgContainer.attr('width') && !!svgContainer.attr('height'));
});

const g = svgContainer.select('g');
test('the <svg> element has <g> children', (t) => {
  t.true(!!g.size());
});

const chart = g.select('g');
test('the <g> element containts another <g>', (t) => {
  t.true(chart.classed('stacked-chart__bar'));
});

test('has barlabels, first entry from barlabels prop', (t) => {
  const barLabels = svgContainer.selectAll('.stacked-chart__barlabel');
  t.true(barLabels.size() === 4);
});

const stackLabelsNodes = svgContainer.selectAll('.stacked-chart__stackedlabel');
test('has stackedlabels', (t) => {
  t.true(stackLabelsNodes.size() === options.data[0].stackLabels.length);
});

test('has a toggle title function', (t) => {
  t.true(typeof sc.toggleTitle === 'function');
});

rollup.config.js
import resolve from 'rollup-plugin-node-resolve';
import babel from 'rollup-plugin-babel';
import butternut from 'rollup-plugin-butternut';
import postcss from 'rollup-plugin-postcss';
import stylus from 'stylus';
import autoprefixer from 'autoprefixer';
import serve from 'rollup-plugin-serve';
import livereload from 'rollup-plugin-livereload';
import cssnano from 'cssnano';
import pkg from './package.json'

const isProd = process.env.NODE_ENV === 'production';

const preprocessor = (content, id) =>
  new Promise((fullfill, reject) => {
    const renderer = stylus(content, {
      filename: id,
      sourcemap: { inline: true }
    });
    renderer.render((err, code) => {
      if (err) {
        return reject(err);
      }
      return fullfill({ code, map: renderer.sourcemap });
    });
  });

const postcssPlugins = [
  autoprefixer
];

if (isProd) {
  postcssPlugins.push(cssnano()); // @FIXME: run cssnano only on the bundle.min.css
}

const plugins = [
  resolve(),
  babel({
    exclude: ['node_modules/**', 'src/styles/main.styl'],
    plugins: ['external-helpers']
  }),
  postcss({
    preprocessor,
    extensions: ['.styl'],
    extract: true,
    plugins: postcssPlugins
  }),
  !isProd && livereload({
    watch: ['examples', 'build']
  }),
  !isProd && serve({
    port: 3000,
    contentBase: ['examples', 'build']
  })
];

const browserBundle = {
  input: 'src/index.js',
  output: [{
    file: pkg.mainmin,
    format: 'umd',
    name: 'OECDCharts',
    sourcemap: true
  }],
  plugins: [
    ...plugins,
    butternut()
  ]
};

const moduleBundle = {
  input: 'src/index.js',
  output: [{
    file: pkg.main,
    format: 'umd',
    name: 'OECDCharts',
    sourcemap: true
  }, {
    file: pkg.module,
    format: 'es',
    sourcemap: true
  }],
  external: isProd ? Object.keys(pkg.dependencies) : [],
  plugins
};

export default isProd ? [browserBundle, moduleBundle] : moduleBundle;


src/modules/OECDChart.js
import { select as d3Select } from 'd3-selection';
import { dispatch as d3Dispatch } from 'd3-dispatch';
import _debounce from 'lodash-es/debounce';

class OECDChart {
  constructor() {
    this.baseDefaultOptions = {};
    this.dispatch = d3Dispatch('infoclick');
    this.on = (eventName, cb) => this.dispatch.on(eventName, cb);

    this.handleInfoClick = this.handleInfoClick.bind(this);
    this.onResize = this.onResize.bind(this);

    const debouncedResize = _debounce(this.onResize, 250);
    window.addEventListener('resize', debouncedResize);
  }

  init(options = {}) {
    this.options = Object.assign({}, this.baseDefaultOptions, this.defaultOptions, options);

    const d3Container = d3Select(this.options.container);

    d3Container
      .classed('oecd-chart__container', true);

    if (this.options.title || this.options.renderInfoButton) {
      const titleContainer =
        d3Container
          .append('div')
          .attr('class', 'oecd-chart__title-container');

      if (this.options.title) {
        titleContainer
          .append('div')
          .attr('class', 'oecd-chart__title')
          .text(this.options.title);
      }

      if (this.options.renderInfoButton) {
        titleContainer
          .append('div')
          .attr('class', 'oecd-chart__info')
          .html('<svg class="oecd-chart__info-icon" width="25" height="25" viewBox="0 0 40 40"><path d="M18.4 15v-3.4h3.3V15h-3.3zM20 33.4q5.5 0 9.4-4t4-9.3-4-9.3-9.4-4-9.4 4-4 9.4 4 9.5 9.4 4zm0-30q7 0 11.8 5t5 11.7-5 12T20 36.7t-11.8-5-5-11.7 5-11.7 11.8-5zm-1.6 25v-10h3.3v10h-3.3z"/></svg>')
          .on('click', this.handleInfoClick);
      }
    }

    this.render(options);
  }

  handleInfoClick() {
    this.dispatch.call('infoclick');
  }

  onResize() {
    const containerWidth = d3Select(this.options.container).node().clientWidth;
    if (containerWidth !== window.innerWidth) this.render(this.options);
  }

  // we call this in render of a specific chart to clear the svg
  // for drawing with new dimensions
  removeSelections(selectors) {
    selectors.forEach(sel => d3Select(this.options.container).selectAll(sel).remove());
  }

  toggleTitle() {
    const titleEl = document.querySelector(`${this.options.container} .oecd-chart__title-container`);
    titleEl.classList.toggle('oecd-chart__title-container--hidden');
  }
}

export default OECDChart;


src/modules/BoxPlot.js
import { select as d3Select } from 'd3-selection';
import { scaleLinear as d3ScaleLinear } from 'd3-scale';
import { axisBottom as d3AxisBottom } from 'd3-axis';

import OECDChart from './OECDChart';

/**
 * A BoxPlot component
 * @example <caption>Browser usage:</caption>
 * const BoxPlotExample = new OECDCharts.BoxPlot({
 *   container: '#BoxPlotExample',
 *   title: 'Box Plot',
 *   extent: [350, 650],
 *   step: 50,
 *   renderInfoButton: true,
 *   data: [
 *    {
 *      values: [480, 500, 530],
 *      colors: ['#fddd5d', '#C7754E', '#900c3f'],
 *      labelLeft: {
 *        text: 'male low'
 *      },
 *      labelRight: {
 *        text: 'male top'
 *      }
 *    },
 *    {
 *      values: [400, 520, 550],
 *      colors: ['#aad356', '#61B77F', '#189aa8']
 *    }
 *   ]
 * });
 * @example <caption>ES6 modules usage:</caption>
 * import { BoxPlot } from 'oecd-simple-charts';
 * import 'oecd-simple-charts/build/oecd-simple-charts.css'
 *
 * const boxPlot = new BoxPlot({ chartOptions });
 *
 * @constructor
 * @param {object}  options - The options object for the Box Plot
 * @param {string}  options.container - The DOM element to use as container
 * @param {string}  options.title - The title to display
 * @param {array}   options.extent - The min and max value for generating the x-axis
 * @param {number}  options.step - Indicates the stepsize for the x-axis ticks
 * @param {string}  options.legend - HTML code for the legend
 * @param {bool}  [options.renderInfoButton = false] - The info-icon for the tooltip, renders after the title
 * @param {int}  [options.fontSize = 12] - The font-size for the labels in px
 * @param {int}  [options.markerHeight = 30] - The height of the marker in px
 * @param {int}  [options.markerHeight = 10] - The width of the marker in px
 * @param {int}  [options.radius = 10] -The radius for the pearl in px
 * @param {array}   options.data - The data as array
 * @param {array}   options.data.values - The values to display
 * @param {array}   options.data.colors - The colors for the elements
 * @param {object}   options.data.labelLeft - (optional) Label for the left marker
 * @param {string}   options.data.labelLeft.label - (optional) Text for the left marker
 * @param {string}   options.data.labelLeft.icon - (optional) Path to icon for the left marker
 * @param {object}   options.data.labelRight - (optional) Label for the right marker
 * @param {string}   options.data.labelRight.label - (optional) Text for the right marker
 * @param {string}   options.data.labelRight.icon - (optional) Path to icon for the right marker
 */
class BoxPlot extends OECDChart {
  constructor(options = {}) {
    super();

    this.defaultOptions = {
      container: null,
      extent: null,
      data: [],
      step: 50,
      legend: null,
      innerMarginTop: 10,
      innerMarginBottom: 10,
      innerMarginLeft: 5,
      innerMarginRight: 5,
      markerHeight: 30,
      markerWidth: 10,
      radius: 10,
      fontSize: 12
    };

    this.init(options);
  }

  render(options) {
    this.update(this.options.data);
  }

  /**
   * @memberof BoxPlot
   * @param {array} data - an array containing objects with the new data
   * @example
   * BoxPlotExample.update([
   *   {
   *     values: [400, 550, 580],
   *     colors: ['#fddd5d', '#C7754E', '#900c3f'],
   *     labelLeft: {
   *       text: 'new label left',
   *     },
   *     labelRight: {
   *       text: 'new label right',
   *     }
   *   },
   *   {
   *     values: [400, 520, 570],
   *     colors: ['#aad356', '#61B77F', '#189aa8']
   *   }
   * ]);
   */
  update(_data) {
    this.options.data = _data;
    const data = BoxPlot.parseData(_data);
    BoxPlot.validateData(data);
    this.getChart(data);

    const markerGroup = this.getMarkerGroups(data);
    this.getMarkers(markerGroup);
    this.getLabels(markerGroup);
  }

  getChart(data) {
    const {
      container,
      extent,
      step,
      legend,
      markerHeight,
      innerMarginTop,
      innerMarginBottom,
      innerMarginLeft,
      innerMarginRight
    } = this.options;

    const d3Container = d3Select(container);
    const dimensions = d3Container.node().getBoundingClientRect();
    const outerWidth = dimensions.width;
    const innerWidth = outerWidth - innerMarginLeft - innerMarginRight;

    this.removeSelections(['.boxplot__svg', '.boxplot__legend']);

    this.x = d3ScaleLinear()
      .range([0, innerWidth - innerMarginRight - innerMarginLeft])
      .domain(extent);
    const height = ((markerHeight + 20) * data.length) + innerMarginTop + innerMarginBottom;

    d3Container
      .append('div')
      .attr('class', 'boxplot__legend')
      .html(legend);

    this.svg = d3Container
      .classed('OECDCharts__BoxPlot', true)
      .append('svg')
      .classed('boxplot__svg', true)
      .attr('width', innerWidth)
      .attr('height', height)
      .append('g')
      .attr('transform', `translate(${innerMarginLeft}, 0)`);

    const ticks = (extent[1] - extent[0]) / step;
    const axis = d3AxisBottom(this.x).ticks(ticks).tickSize(-height);

    this.svg.append('g')
      .classed('boxplot__axis', true)
      .attr('transform', `translate(${innerMarginLeft}, ${height - innerMarginTop - innerMarginBottom})`)
      .call(axis);
  }

  getMarkerGroups(_data) {
    const markerGroup = this.svg.selectAll('.boxplot__marker-group')
      .data(_data, d => JSON.stringify(d));

    markerGroup.exit().remove();

    return markerGroup.enter()
      .append('g')
      .classed('boxplot__marker-group', true)
      .attr('transform', (d, i) => `translate(0, ${((this.options.markerHeight + 20) * i) + 1})`);
  }

  getMarkers(markerGroup) {
    markerGroup.append('line')
      .classed('marker-group__line', true)
      .attr('x1', d => this.x(d.values[0]) + 5)
      .attr('x2', d => this.x(d.values[2]) + 5)
      .attr('y1', this.options.markerHeight / 2)
      .attr('y2', this.options.markerHeight / 2)
      .style('stroke', d => d.colors[1]);

    markerGroup.selectAll('.marker-group__rect')
      .data(d => [0, 2].map(i => ({ value: d.values[i], color: d.colors[i] })))
      .enter()
      .append('rect')
      .classed('marker-group__rect', true)
      .attr('height', this.options.markerHeight)
      .attr('width', this.options.markerWidth)
      .attr('x', d => (this.x(d.value) + 5) - (this.options.markerWidth / 2))
      .style('fill', d => d.color);

    markerGroup.selectAll('.marker-group__circle')
      .data(d => [d])
      .enter()
      .append('circle')
      .classed('marker-group__circle', true)
      .attr('r', this.options.radius)
      .attr('cx', d => this.x(d.values[1]) + 5)
      .attr('cy', this.options.markerHeight / 2)
      .style('fill', d => d.colors[1]);
  }

  getLabels(markerGroup) {
    this.getLabel(markerGroup, 'left');
    this.getLabel(markerGroup, 'right');
  }

  getLabel(markerGroup, pos) {
    const left = pos === 'left';
    const label = markerGroup
      .filter(d => (left ? d.labelLeft : d.labelRight))
      .selectAll(`.marker-group__label--${pos}`)
      .data(d => [d])
      .enter()
      .append('g')
      .classed(`marker-group__label--${pos}`, true);

    // add label text
    label
      .filter(d => (left ? d.labelLeft.text : d.labelRight.text))
      .selectAll(`.marker-group__label--text-${pos}`)
      .data((d) => {
        const { text, icon } = (left ? d.labelLeft : d.labelRight);
        return [{ text, icon, value: (left ? d.values[0] : d.values[2]) }];
      })
      .enter()
      .append('text')
      .classed(`marker-group__label--text-${pos}`, true)
      .attr('y', (this.options.markerHeight / 2) + (this.options.fontSize / 4))
      .attr('x', (d) => {
        const offset = d.icon ? 30 : 5;
        return left ? this.x(d.value) - (offset) : this.x(d.value) + offset + 10;
      })
      .text(d => d.text)
      .attr('text-anchor', left ? 'end' : 'start')
      .attr('font-size', this.options.fontSize);

    // add label icon
    label
      .filter(d => (left ? d.labelLeft.icon : d.labelRight.icon))
      .selectAll(`.marker-group__label--icon-${pos}`)
      .data((d) => {
        const { icon } = (left ? d.labelLeft : d.labelRight);
        return [{ icon, value: (left ? d.values[0] : d.values[2]) }];
      })
      .enter()
      .append('svg:image')
      .classed(`marker-group__label--icon-${pos}`, true)
      .attr('y', (this.options.markerHeight / 2) - 10)
      .attr('x', d => (left ? this.x(d.value) - 25 : this.x(d.value) + 15))
      .attr('xlink:href', d => d.icon)
      .attr('width', 20)
      .attr('height', 20);


    BoxPlot.arrangeLabels(markerGroup, this.svg);
  }

  static arrangeLabels(container, svgContainer) {
    const svgBB = svgContainer.node().getBoundingClientRect();

    container.selectAll('.marker-group__label--left')
      .each((node, i, nodes) => {
        const labelNode = nodes[i];
        const bb = labelNode.getBoundingClientRect();

        if (bb.left <= svgBB.left) {
          d3Select(labelNode)
              .attr('transform', `translate(${bb.width + 20}, ${bb.height / 1.1})`);
        }
      });

    container.selectAll('.marker-group__label--right')
      .each((node, i, nodes) => {
        const labelNode = nodes[i];
        const bb = labelNode.getBoundingClientRect();

        if (bb.right >= svgBB.right) {
          d3Select(labelNode)
            .attr('transform', `translate(-${bb.width + 20}, ${bb.height / 1.1})`);
        }
      });
  }

  static parseData(_data) {
    return _data.map((d) => {
      d.colors = d.colors || ['#f0f0f0', '#555', '#000'];
      return d;
    });
  }

  static validateData(_data) {
    const invalidData = _data.filter(
      d => (d.values && d.values.length !== 3)
    );

    if (invalidData.length) {
      throw Error('invalid data: values needs three items.');
    }
  }
}

export default BoxPlot;
